lwm/lwm.man
.\" lwm, a window manager for X11
.\" Copyright (C) 1997-2003 Elliott Hughes, James Carter
.\" 
.\" This program is free software; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License
.\" as published by the Free Software Foundation; either version 2
.\" of the License, or (at your option) any later version.
.\" 
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\" 
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
.\" 
.TH LWM 1
.SH NAME
lwm \- Lightweight Window Manager for the X Window System
.SH SYNTAX
\fBlwm \fP[ \fB\-s\fP \fIsession-id\fP ] 
.SH DESCRIPTION
\fILwm\fP is a window manager for the X Window System. It provides enough
features to allow the user to manage their windows, and no more.
.PP
Windows are surrounded by a frame with a 
titlebar at the top next to a small box. The frame is a grey colour for
all windows except that which has the input focus, where it is black.
.PP
In the default configuration, \fIlwm\fP uses the enter-to-focus scheme, where
moving the pointer into a window gives that window the input focus.
\fILwm\fP may also be configured to use the click-to-focus scheme, where a
window must be clicked on (with any button) to receive the input focus. Clicking
on a window in this mode causes the window to be raised. Note that a click
used to focus a window is always swallowed by \fIlwm\fP, so clicking a
button in a new window requires two clicks.
.PP
A button 1 click on a window frame brings that window to the top. Dragging
button 1 on the frame of a resizable window repositions that edge of
the window. If a corner rather than an edge is dragged, then both edges
forming the corner are repositioned. While you're reshaping a window,
a little window pops up to show you the window's current size.
.PP
In the default configuration, button 1 on the root window does nothing.
.PP
Button 2 is used to drag a window by its frame, repositioning the window
but maintaining its position in the window stack.
.PP
In the default configuration, button 2 on the root window brings up a
new shell.
.PP
A button 3 click on a window frame hides that window.  Pressing
button 3 on the root window brings up a menu. The various items unhide
the named hidden window.
.PP
A button 3 click in the frame while Shift is held down pushes the window
to the back, under any other windows. (Users with 4-button mice are
encouraged to use their fourth button for this function.)
.PP
A click with any button inside the little white box in a window's frame
can be used to close the window.
.SH OPTIONS
\fILwm\fP accepts the following command line options:
.PP
.TP 8
.B \-s
specifies a client ID for the X Session Management system, and is used
exclusively by session managers.
.SH RESOURCES
\fILwm\fP understands the following X resources:
.TP 12
.B titlefont
font used in window titles
.TP 12
.B popupFont
font used in popup window (menu/size indicator)
.TP 12
.B border
width in pixels of window borders
.TP 12
.B button1
program spawned when button 1 is clicked on the root window
.TP 12
.B button2
program spawned when button 2 is clicked on the root window
.TP 12
.B focus
focus mode, one of "enter" for enter-to-focus (or sloppy focus), or
"click" for click-to-focus
.SH "SEE ALSO"
.PP
X(7)
.SH AUTHORS
Elliott Hughes <ehughes@bluearc.com>,
James Carter <james@jfc.org.uk>


lwm/ChangeLog
Change Log for "lwm"

2009-11-20	jfc	York
	Improved performance by only checking for pending X events when the
	socket it ready for reading.

	Fixed applyGravity() bug that caused frameless windows to be
	mis-positioned.

	Applied a workaround in destroy() to avoid error reports when closing
	windows.

2005-01-28	jfc	York

	Applied a patch from Chris Reece that ensures that the popup
	menu does not disappear off the bottom of the screen.

2004-09-30	jfc	York

	Fixed an issue with IRIX 6.5 and lwm, where the root menu could
	not be used because motion events had coordinates with respect to
	the popup, not the root. Fixed by explicitly using the root window
	coordinates.

	Released lwm-1.2.1.

2004-09-28	jfc	York

	Added missing -lSM to no_xmkmf_makefile.

2003-12-09	jfc	York

	Fixed bug (reported by Matthew Wilcox) where windows with extremely
	long names could cause the pop menu to be unusable. Fixed by
	maintaining a separate, shortened name for the menu, if
	necessary ("this is a very very [...] ry long window name"). This
	takes no account of UTF-8 names as yet.

2003-12-08	jfc	York

	Fixed bug (reported by Eugene Wong) where resizing the top of
	a window would cause it to jump up several pixels. The height of
	the titlebar was not being considered when calculating mouse
	motion in reshaping_motionnotify().

	Modified manage() to avoid autoplacing windows during
	initialisation.

	Released lwm-1.2.0.

2003-12-03	jfc	York

	Applied patch from Elliott that gives focus to new windows in
	click-to-focus mode.

	Changed the buttonpress code in disp.c to ignore scroll wheel
	"clicks".

	Modified Client_Remove so that, in click-to-focus mode, it
	refocuses on the most sensible window (either the top window,
	or the window that the closing window was a transient for).

	Attempted to fix the edit placement bug by adding titleHeight()
	to the supplied X coordinate during a ConfigureRequest event,
	and not attempting to fix clients that don't supply a border
	width during a configure request.

	Fixed fullscreen-mode bug where galeon windows appeared to jump
	up and to the left after the first click.

	Released lwm-1.1.7.

2003-11-28	jfc	York

	Changed the behaviour when unhiding a window in click-to-focus mode.
	An unhidden window now automatically gets focus in this mode.

	Fixed a small bug in the session management code that could
	cause a crash when lwm quit.

	Moved a call to ewmh_set_client_list make before ewmh was
	initialised.

	Changed lwm's behaviour when minimising windows. Button three must
	now be pressed and released before the window is hidden (or
	moved to the bottom of the stack). This ensures that lwm swallows
	all the events generated during the operation, and allows the user
	to back out of the operation by moving the mouse out of the window
	before releasing the button.

	Removed include of Xm/MwmUtil.h in manage.c, and the HAVE_MOTIF
	kludge from the Imakefile, in favour of copying the few lines
	that are required from Xm/MwmUtil.h (LessTif, so hopefully no
	licensing issues).

	Fixed bug that caused the last cursor displayed in a frame
	to be incorrectly used when moving into the frame when the
	root menu was on screen. This is done by brute force -
	see Client_ResetAllCursors().

	Released lwm-1.1.6.

2003-11-26	jfc	York

	Fixed bug in Client_MakeSane that caused occasional crashes
	during window moves/resizing. Should investigate why it occasionally
	gets called with a NULL client.

2003-11-03	jfc	York

	Added an entry for LeaveNotify in the dispatch table (disp.c).

2003-08-13	jfc	York

	Removed "error" message when lwm fails to connect to a session
	manager. This isn't actually an error and the message is confusing.

2003-08-01	jfc	York

	Fixed bug that allowed clients to grab the focus and confuse lwm.

	Cleaned up the code for raising and lowering clients, and added
	code to prevent a client from being raised above its transients.

	Retired disp.old, and CLOSE_PATCH.txt.

	Added an edge resistance to the workarea, so that window may
	be moved to the edge of the workarea without precise mousing,
	as requested by MAD. EDGE_RESIST in lwm.h defines the number of
	pixels of resistance and may be safely set to zero.

	Released lwm-1.1.5.

2003-07-31	jfc	York

	In click-to-focus mode, always draw the box in the frame.

2003-07-29	jfc	York

	Added a click-to-focus mode. The default remains (sloppy)
	enter-to-focus.

	Released lwm-1.1.4.


2003-07-28	jfc	York

	Updated no_xmkmf_makefile to reflect the changes made since 1.01.

2003-07-10	jfc	York

	Fixed a bug in manage.c than prevented lwm compiling on systems
	with no variety of Motif installed. If this means you, remove
	_DHAVE_MOTIF from Imakefile.

	Released lwm-1.1.3.

2003-07-08	jfc	York

	Added support for NET_MOVERESIZE, but I cannot find any
	applications the want to use it, apart from the keyboard
	variants. I don't know what to do about the keyboard move/resize.

2003-07-03	jfc	York

	Fixed a few buglets thrown up by running lwm through the compiler
	with all warnings on.

2003-07-02	jfc	York

	In Client_MakeSane(), added a check to prevent windows being
	moved into a position where they might be completely obscured
	by panels/docks.

	Changed ewmh_set_strut() to run Client_MakeSane() across all
	clients when the work area changes. This avoids clients getting
	lost behind panels/docks.

	Added support for _NET_WM_STATE_ABOVE and
	_NET_WM_STATE_BELOW. Added fix_stack() to maintain the window
	stack as dictated by the EWMH spec.

2003-07-01	jfc	York

	Added support for _NET_WM_STRUT. lwm now maintains _NET_WORKAREA
	correctly, and takes the reserved space into account in its
	window placing algorithm.

	Released lwm-1.1.2.

2003-06-30	jfc	York

	Fixed bug that caused tk menus to be badly placed placed by
	sending a configure notify where appropriate in setactive().

	Removed compile time option of prepending window title's with
	the client machines's name (PREPEND_CLIENT_MACHINE).


lwm/lwm.c
/*
 * lwm, a window manager for X11
 * Copyright (C) 1997-2003 Elliott Hughes, James Carter
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <locale.h>
#include <errno.h>

#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>

#include <signal.h>

#include <X11/X.h>
#include <X11/Xos.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>

#include "lwm.h"

Mode mode;			/* The window manager's mode. (See "lwm.h".) */
int start_x;			/* The X position where the mode changed. */
int start_y;			/* The Y position where the mode changed. */

Display * dpy;			/* The connection to the X server. */
int screen_count;		/* The number of screens. */
ScreenInfo * screens;		/* Information about these screens. */
ScreenInfo * current_screen;

XFontSet font_set = NULL;	/* Font set for title var */
XFontSetExtents *font_set_ext = NULL;
XFontSet popup_font_set = NULL;	/* Font set for popups */
XFontSetExtents *popup_font_set_ext = NULL;

Bool shape;			/* Does server have Shape Window extension? */
int shape_event;		/* ShapeEvent event type. */

/* Atoms we're interested in. See the ICCCM for more information. */
Atom wm_state;
Atom wm_change_state;
Atom wm_protocols;
Atom wm_delete;
Atom wm_take_focus;
Atom wm_colormaps;
Atom compound_text;

/** Netscape uses this to give information about the URL it's displaying. */
Atom _mozilla_url;

/*
 * if we're really short of a clue we might look at motif hints, and
 * we're not going to link with motif, so we'll have to do it by hand
 */
Atom motif_wm_hints;

char *argv0;

static void initScreens(void);
static void initScreen(int);

/*ARGSUSED*/
extern int
main(int argc, char *argv[]) {
	XEvent ev;
	struct sigaction sa;
	int dpy_fd, max_fd;

	argv0 = argv[0];

	mode = wm_initialising;

	setlocale(LC_ALL,"");

	/* Open a connection to the X server. */
	dpy = XOpenDisplay(NULL);
	if (dpy == 0)
		panic("can't open display.");

	parseResources();

	/* Set up an error handler. */
	XSetErrorHandler(errorHandler);

	/* Set up signal handlers. */
	signal(SIGTERM, Terminate);
	signal(SIGINT, Terminate);
	signal(SIGHUP, Terminate);

	/* Ignore SIGCHLD. */
	sa.sa_handler = SIG_IGN;
#ifdef SA_NOCLDWAIT
	sa.sa_flags = SA_NOCLDWAIT;
#else
	sa.sa_flags = 0;
#endif
	sigemptyset(&sa.sa_mask);
	sigaction(SIGCHLD, &sa, 0);

	/* Internalize useful atoms. */
	wm_state = XInternAtom(dpy, "WM_STATE", False);
	wm_change_state = XInternAtom(dpy, "WM_CHANGE_STATE", False);
	wm_protocols = XInternAtom(dpy, "WM_PROTOCOLS", False);
	wm_delete = XInternAtom(dpy, "WM_DELETE_WINDOW", False);
	wm_take_focus = XInternAtom(dpy, "WM_TAKE_FOCUS", False);
	wm_colormaps = XInternAtom(dpy, "WM_COLORMAP_WINDOWS", False);
	compound_text = XInternAtom(dpy, "COMPOUND_TEXT", False);
	
	_mozilla_url = XInternAtom(dpy, "_MOZILLA_URL", False);

	motif_wm_hints = XInternAtom(dpy, "_MOTIF_WM_HINTS", False);

	ewmh_init();
	
	/*
	 * Get fonts for our titlebar and our popup window. We try to
	 * get Lucida, but if we can't we make do with fixed because everyone
	 * has that.
	 */
	{
		/* FIXME: do these need to be freed? */
		char **missing;
		char *def;
		int missing_count;

		font_set = XCreateFontSet(dpy, font_name,
			&missing, &missing_count, &def);
		if (font_set == NULL)
			font_set = XCreateFontSet(dpy, "fixed",
				&missing, &missing_count, &def);
		if (font_set == NULL)
			panic("unable to create font set for title font");
		if (missing_count > 0)
			fprintf(stderr,"%s: warning: missing %d charset"
				"%s for title font\n", argv0, missing_count,
				(missing_count == 1)?"":"s");
		font_set_ext = XExtentsOfFontSet(font_set);

		popup_font_set = XCreateFontSet(dpy, popup_font_name,
			&missing, &missing_count, &def);
		if (popup_font_set == NULL)
			popup_font_set = XCreateFontSet(dpy, "fixed",
				&missing, &missing_count, &def);
		if (popup_font_set == NULL)
			panic("unable to create font set for popup font");
		if (missing_count > 0)
			fprintf(stderr,"%s: warning: missing %d charset"
				"%s for popup font\n", argv0, missing_count,
				(missing_count == 1)?"":"s");
		popup_font_set_ext = XExtentsOfFontSet(popup_font_set);
	}
	
	initScreens();
	ewmh_init_screens();
	session_init(argc, argv);
	
	/* See if the server has the Shape Window extension. */
	shape = serverSupportsShapes();
	
	/*
	 * Initialisation is finished, but we start off not interacting with the
	 * user.
	 */
	mode = wm_idle;
	
	/*
	 * The main event loop.
	 */
	dpy_fd = ConnectionNumber(dpy);
	max_fd = dpy_fd + 1;
	if (ice_fd > dpy_fd) max_fd = ice_fd + 1;
	for (;;) {
		fd_set readfds;

		FD_ZERO(&readfds);
		FD_SET(dpy_fd, &readfds);
		if (ice_fd > 0) FD_SET(ice_fd, &readfds);
		if (select(max_fd, &readfds, NULL, NULL, NULL) > -1) {
		    if (FD_ISSET(dpy_fd, &readfds)) {
			    while (XPending(dpy)) {
				XNextEvent(dpy, &ev);
				dispatch(&ev);
			    }
		    }
		    if (ice_fd > 0 && FD_ISSET(ice_fd, &readfds)) {
			    session_process();
		    }
		}
	}
}

void
sendConfigureNotify(Client *c) {
	XConfigureEvent ce;

	ce.type = ConfigureNotify;
	ce.event = c->window;
	ce.window = c->window;
	if (c->framed == True) {
		ce.x = c->size.x + border;
		ce.y = c->size.y + border;
		ce.width = c->size.width - 2 * border;
		ce.height = c->size.height - 2 * border;
		ce.border_width = c->border;
	} else {
		ce.x = c->size.x;
		ce.y = c->size.y;
		ce.width = c->size.width;
		ce.height = c->size.height;
		ce.border_width = c->border;
	}
	ce.above = None;
	ce.override_redirect = 0;
	XSendEvent(dpy, c->window, False, StructureNotifyMask, (XEvent *) &ce);
}

extern void
scanWindowTree(int screen) {
	unsigned int i;
	unsigned int nwins;
	Client * c;
	Window dw1;
	Window dw2;
	Window * wins;
	XWindowAttributes attr;
	
	XQueryTree(dpy, screens[screen].root, &dw1, &dw2, &wins, &nwins);
	for (i = 0; i < nwins; i++) {
		XGetWindowAttributes(dpy, wins[i], &attr);
		if (attr.override_redirect /*|| isShaped(wins[i])*/ || wins[i] == screens[screen].popup)
			continue;
		c = Client_Add(wins[i], screens[screen].root);
		if (c != 0 && c->window == wins[i]) {
			c->screen = &screens[screen];
			c->size.x = attr.x;
			c->size.y = attr.y;
/* we'll leave it until it's managed
			if (c->framed == True) {
				c->size.x -= border;
				c->size.y -= border;
			}
*/
			c->size.width  = attr.width;
			c->size.height = attr.height;
/* we'll leave it until it's managed
			if (c->framed == True) {
				c->size.width  += 2 * border;
				c->size.height += 2 * border;
			}
*/
			c->border = attr.border_width;
			if (attr.map_state == IsViewable) {
				c->internal_state = IPendingReparenting;
				manage(c, 1);
			}
		}
	}
	XFree(wins);
}

/*ARGSUSED*/
extern void
shell(ScreenInfo * screen, int button, int x, int y) {
	char * command = NULL;
	char * sh;
	
	/* Get the command we're to execute. Give up if there isn't one. */
	if (button == Button1)
		command = btn1_command;
	if (button == Button2)
		command = btn2_command;
	if (command == NULL)
		return;
	
	sh = getenv("SHELL");
	if (sh == 0)
		sh = "/bin/sh";
	
	switch (fork()) {
	case 0:		/* Child. */
		close(ConnectionNumber(dpy));
		if (screen && screen->display_spec != 0)
			putenv(screen->display_spec);
		execl(sh, sh, "-c", command, NULL);
		fprintf(stderr, "%s: can't exec \"%s -c %s\"\n", argv0, sh,
			command);
		execlp("xterm", "xterm", NULL);
		exit(EXIT_FAILURE);
	case -1:	/* Error. */
		fprintf(stderr, "%s: couldn't fork\n", argv0);
		break;
	}
}

extern int
titleHeight(void) {
	return font_set_ext->max_logical_extent.height;
}

extern int
ascent(XFontSetExtents *font_set_ext) {
	return abs(font_set_ext->max_logical_extent.y);
}

extern int
popupHeight(void) {
	return popup_font_set_ext->max_logical_extent.height;
}

extern int
titleWidth(XFontSet font_set, Client *c) {
	XRectangle ink;
	XRectangle logical;
	char *name;
	int namelen;

	if (c == NULL) return 0;
	if (c->menu_name == NULL) {
		name = c->name;
		namelen = c->namelen;
	} else {
		name = c->menu_name;
		namelen = c->menu_namelen;
	}
	if (name == NULL) return 0;
#ifdef X_HAVE_UTF8_STRING
	if (c->name_utf8 == True) 
		Xutf8TextExtents(font_set, name, namelen,
			&ink, &logical);
	else
#endif
		XmbTextExtents(font_set, name, namelen,
			&ink, &logical);

	return logical.width;
}

extern int
popupWidth(char *string, int string_length) {
	XRectangle ink;
	XRectangle logical;

	XmbTextExtents(popup_font_set, string, string_length,
		&ink, &logical);

	return logical.width;
}

static void
initScreens(void) {
	int screen;
	
	/* Find out how many screens we've got, and allocate space for their info. */
	screen_count = ScreenCount(dpy);
	screens = (ScreenInfo *) malloc(screen_count * sizeof(ScreenInfo));
	
	/* Go through the screens one-by-one, initialising them. */
	for (screen = 0; screen < screen_count; screen++) {
		initialiseCursors(screen);
		initScreen(screen);
		scanWindowTree(screen);
	}
}

static void
initScreen(int screen) {
	XGCValues gv;
	XSetWindowAttributes attr;
	XColor colour, exact;
	int len;
	char * display_string = DisplayString(dpy);
	char * colon = strrchr(display_string, ':');
	char * dot = strrchr(display_string, '.');
	
	/* Set the DISPLAY specification. */
	if (colon) {
		len = 9 + strlen(display_string) + ((dot == 0) ? 2 : 0) + 10;
		screens[screen].display_spec = (char *) malloc(len);
		sprintf(screens[screen].display_spec, "DISPLAY=%s", display_string);
		if (dot == 0) dot = screens[screen].display_spec + len - 3;
		else dot = strrchr(screens[screen].display_spec, '.');
		sprintf(dot, ".%i", screen);
	} else {
		screens[screen].display_spec = 0;
	}

	/* Find the root window. */
	screens[screen].root = RootWindow(dpy, screen);
	screens[screen].display_width = DisplayWidth(dpy, screen);
	screens[screen].display_height = DisplayHeight(dpy, screen);
	screens[screen].strut.left = 0;
	screens[screen].strut.right = 0;
	screens[screen].strut.top = 0;
	screens[screen].strut.bottom = 0;
	
	/* Get the pixel values of the only two colours we use. */
	screens[screen].black = BlackPixel(dpy, screen);
	screens[screen].white = WhitePixel(dpy, screen);
	XAllocNamedColor(dpy, DefaultColormap(dpy, screen), "DimGray", &colour, &exact);
	screens[screen].gray = colour.pixel;
	
	/* Set up root (frame) GC's. */
	gv.foreground = screens[screen].black ^ screens[screen].white;
	gv.background = screens[screen].white;
	gv.function = GXxor;
	gv.line_width = 1;
	gv.subwindow_mode = IncludeInferiors;
	screens[screen].gc_thin = XCreateGC(dpy, screens[screen].root,
		GCForeground | GCBackground | GCFunction |
		GCLineWidth | GCSubwindowMode, &gv);
	
	gv.line_width = 2;
	screens[screen].gc = XCreateGC(dpy, screens[screen].root,
		GCForeground | GCBackground | GCFunction |
		GCLineWidth | GCSubwindowMode, &gv);
	
	/* Create a window for our popup. */
	screens[screen].popup = XCreateSimpleWindow(dpy, screens[screen].root,
		0, 0, 1, 1, 1, screens[screen].black, screens[screen].white);
	attr.event_mask = ButtonMask | ButtonMotionMask | ExposureMask;
	XChangeWindowAttributes(dpy, screens[screen].popup, CWEventMask, &attr);
	
	/* Create menu GC. */
	gv.line_width = 1;
	screens[screen].menu_gc = XCreateGC(dpy, screens[screen].popup,
		GCForeground | GCBackground | GCFunction |
		GCLineWidth | GCSubwindowMode, &gv);
	
	/* Create size indicator GC. */
	gv.foreground = screens[screen].black;
	gv.function = GXcopy;
	screens[screen].size_gc = XCreateGC(dpy, screens[screen].popup,
		GCForeground | GCBackground | GCFunction |
		GCLineWidth | GCSubwindowMode, &gv);
	
	/* Announce our interest in the root window. */
	attr.cursor = screens[screen].root_cursor;
	attr.event_mask = SubstructureRedirectMask | SubstructureNotifyMask |
		ColormapChangeMask | ButtonPressMask | PropertyChangeMask |
		EnterWindowMask;
	XChangeWindowAttributes(dpy, screens[screen].root, CWCursor |
		CWEventMask, &attr);
	
	/* Make sure all our communication to the server got through. */
	XSync(dpy, False);
}

/**
Find the screen for which root is the root window.
*/
ScreenInfo *
getScreenFromRoot(Window root) {
	int screen;
	
	for (screen = 0; screen < screen_count; screen++)
		if (screens[screen].root == root)
			return &screens[screen];
	
	return 0;
}
